vo_gpu_next: refactor subtitle rendering

haasn opened the pull request:
Render subs at the video resolution, rather than the overlay resolution.
Uses the new APIs found in libplacebo 196+, to allow controlling the OSD
resolution even for image-attached overlays.

Also fixes an issue where the overlay state did not get correctly
updated while paused. To avoid regenerating the OSD / flushing the cache
constantly, we keep track of OSD changes and only regenerate the OSD
when the OSD state is expected to change in some way (e.g. resolution
change). This requires introducing a new VOCTRL to inform the VO when
the UPDATE_OSD-tagged options have changed.

Fixes #9744, #9524, #9399 and #9398.

TODO:
- [x] video-rotate etc. breaks subtitles entirely
- [x] test to see if it interacts as expected with sub-use-margins?
- [x] should we support `--blend-subtitles=video` still? (it won't blend it onto the video, though, just generate it at video res - with bilinear scaling...)

(*should* subtitles be rotated along with the video, or not?)

Read this before you submit this pull request:
https://github.com/mpv-player/mpv/blob/master/DOCS/contribute.md

Reading this link and following the rules will get your pull request reviewed
and merged faster. Nobody wants lazy pull requests.

haasn commented:
Hmm, as written this won't work with `sub-use-margins`, because libplacebo creates mix frames at the resolution of the dst rect, not the full frame.

We could recover the `vo_gpu` behavior by re-introducing `--blend-subtitles=yes/no`, the latter of which will merge subtitles into the OSD rendering, as currently.

Akemi commented:
since i use _blend-subtitles=yes_, so i would appreciate the reintroduction of that option, even though it doesn't affect mit directly atm.

what is the difference between "Render subs at the video resolution, rather than the overlay resolution." and "--blend-subtitles=video"? asking because of the TODO and was wondering.

will this still work properly with interpolation (when it is reintroduced in gpu-next if it isn't yet) as mentioned in this commit https://github.com/mpv-player/mpv/commit/8c43e12b20a52dc61b53bc3eec779aa07b8c8220, eg are/would subs be interpolated too?

haasn commented:
> since i use blend-subtitles=yes, so i would appreciate the reintroduction of that option, even though it doesn't affect mit directly atm.

To clarify, `blend-subtitles=yes` is the current default behavior. My comment was about a need to reintroduce `blend-subtitles=no` (for subtitles inside margins). Though something else to consider would be, instead of tying that to blend-subtitles, to simply add an option to `pl_renderer` to allow compositing the full framebuffer (including subtitles). I actually like that approach more, I think.

(And the open question remains about whether or not `blend-subtitles=video` is important enough behavior to reintroduce, even if it will mean low-quality bilinear scaling for the subtitles)

> what is the difference between "Render subs at the video resolution, rather than the overlay resolution." and "--blend-subtitles=video"? asking because of the TODO and was wondering.

None. I'm not sure whence you got that there was a distinction

> will this still work properly with interpolation (when it is reintroduced in gpu-next if it isn't yet) as mentioned in this commit 8c43e12, eg are/would subs be interpolated too?

Yeah, it already does. Except if we reimplement `blend-subtitles=no` in the stupid way. (Though not in the way I described in this post)

Akemi commented:
thanks for the clarification.

> None. I'm not sure whence you got that there was a distinction

i kinda got the impression they were two different things, since "Render subs at the video resolution, rather than the overlay resolution." is from your commit message but the TODO mentions "should we support --blend-subtitles=video still?". i think i kind misunderstood the the intention behind the TODO.

`blend-subtitles=video` could be important since it has a bit less of a performance/resource impact, especially in comparison to subtitles rendered at the upscaled video frame resolution/high res displays. i believe we had some reported issues in the past where this helped.

dyphire commented:
I prefer `blend-subtitles=no`. So it's better for me to reintroduce it.

If `sub-use-margins` doesn't work for now. So `sub-ass-force-margins` should not work either.

haasn commented:
@Akemi oh, I see; I have the order screwed up in the commit message. My bad.

Obegg commented:
If I may add my opinion:
```blend-subtitles=no``` is the way to go, since it shows the subtitles as they were intended to be seen.
Subtitles, especially ```.ass```, already have their own styling and have their own blending to the video (colors and text movements).
Adding additional blending (```blend-subtitles=yes```) will cause a color shift.

https://mpv.io/manual/master/#options-blend-subtitles

![image](https://user-images.githubusercontent.com/13978119/154136011-fec8a2b4-0975-4830-aa23-f05e07e91373.png)

haasn commented:
@Obegg that paragraph is irrelevant to gpu-next, which always treats subtitles as in the video color space

dyphire commented:
> @Obegg that paragraph is irrelevant to gpu-next, which always treats subtitles as in the video color space (except for HDR)

Does this mean that the option of sub-ass-vsfilter-color-compat will not work?

haasn commented:
So, updated and fixed the rotation-related issues. Here's a summary of the current state:

1. This basically forces `--blend-subtitles=yes` to be the new behavior (as opposed to the old forced behavior of `--blend-subtitles=video`)
2. `--blend-subtitles`, as of this commit, continues being ignored
3. Subtitles cannot be rendered into the black bar

I decided to leave this commit in this state because as-is it represents a strict improvement, and I'm not sure which of the many colors to paint the blend-subs bikeshed.

haasn commented:
Also, the `--blend-subtitles` documentation is basically completely wrong for `gpu-next`. I'll probably ignore this for the time being until `gpu-next` becomes `gpu`. But maybe I should start a wiki page of known gpu/gpu-next differences somewhere?

Obegg commented:
Just a question - Is it possible that ```vo=gpu-next``` renders subtitles using shaders?
Asking this because when I use ```vo=gpu-next``` and use .ass subtitles and a scene comes up where there are a lot of signs and text on the screen = mpv is dropping frames, so I'm asking to know if this is the intended behavior and if I should open a issue (which by the way I don't think it will change much since the dropped frames are not showing on logs, right?)

By the way I've noticed my GPU is doing all the hard work and use 100%, so is it possible to use 100% CPU somehow? (I have 13th gen Intel CPU and I want to see what's better at rendering, my GPU or my CPU)
